### PR TITLE
General fixes to import/export

### PR DESCRIPTION
**Pull request by GitHK**
## What do these changes do?

- [x] quality was not saved while importing
- [x] duplicate study API https://github.com/ITISFoundation/osparc-issues/issues/402
- [x] archiving module functions interface updated
- [x] unarchiving will no longer create race conditions

**Note:** proper locking will come with separate PR, as a lot of systems need to be touched in refactoring the status.

## Related issue/s

closes https://github.com/ITISFoundation/osparc-issues/issues/402

### CODE REVIEW

File: api/specs/webserver/openapi-projects.yaml [hunk 1]
```diff
@@ -222,6 +217,31 @@ paths:
         default:
           $ref: "#/components/responses/DefaultErrorResponse"
 
+  /projects/{project_id}/duplicate:
```
Comment (odeimaiz) [inline api/specs/webserver/openapi-projects.yaml:220]: :duplicate maybe?

**Review by odeimaiz — COMMENTED** [review]
I miss the "Exporting" and "Duplicating" states

File: api/specs/webserver/openapi-projects.yaml (hunk 1 above)
Comment (GitHK) [reply api/specs/webserver/openapi-projects.yaml:220]: no, in this file all the routes are like this.

**Comment by GitHK** [comment]
> I miss the "Exporting" and "Duplicating" states

Will come with future PR, I wanted to keep them separated for now

File: services/web/server/tests/integration/test_exporter.py [hunk 2]
```diff
@@ -335,15 +349,45 @@ async def test_import_export_import(
...
 
+    # duplicate newly imported project
+    url_duplicate = client.app.router["duplicate_project"].url_for(
+        project_id=imported_project_uuid
+    )
+    assert url_duplicate == URL(
+        API_PREFIX + f"/projects/{imported_project_uuid}:duplicate"
+    )
+    async with await client.post(url_duplicate, timeout=10) as duplicate_response:
+        assert duplicate_response.status == 200, await duplicate_response.text()
```
Comment (pcrespov) [inline services/web/server/tests/integration/test_exporter.py:360]: FYI: check ``assert_status`` in [pytest_simcore helpers](https://github.com/ITISFoundation/osparc-simcore/blob/2a16137461/packages/pytest-simcore/src/pytest_simcore/helpers/utils_assert.py#L12)

File: services/web/server/tests/integration/test_exporter.py [hunk 3]
```diff
@@ -287,7 +297,8 @@ async def import_study_from_file(client, file_path: Path) -> str:
 
 
 @pytest.mark.parametrize("export_version", get_exported_projects())
-async def test_import_export_import(
+async def test_import_export_import_duplicate(
+    loop,
```
Comment (pcrespov) [inline services/web/server/tests/integration/test_exporter.py:301]: Should you check that s3 data was also duplicated?

**Review by pcrespov — APPROVED** [review]
Nice! Now we can duplicate and speed-zip/unzip ! :-)

Please consider my comments

File: services/web/server/src/simcore_service_webserver/exporter/archiving.py [hunk 4]
```diff
@@ -14,12 +14,116 @@
...
+    }
+    # extract all possible subdirectories
+    subdirectories = {x.parent for x in full_destination_paths}
+    # create all subdirectories before extracting
+    for subdirectory in subdirectories:
+        Path(subdirectory).mkdir(parents=True, exist_ok=True)
+
+
+async def unarchive_dir(archive_to_extract: Path, destination_folder: Path) -> None:
+    # running in process poll is not ideal for concurrency issues
```
[the remark this answers is not in the document]
Comment (GitHK) [reply services/web/server/src/simcore_service_webserver/exporter/archiving.py:52]: Moved the comment where it made more sense. Yes I am doing a parallel extraction, but I've removed the possibility of having a race condition when extracting a file from the zip archive by creating the entire directory structure before hand.

File: services/web/server/src/simcore_service_webserver/exporter/request_handlers.py [hunk 5]
```diff
@@ -88,4 +88,37 @@ async def import_project(request: web.Request):
...
+        # return the duplicated study ID
+        duplicated_project_uuid = await study_duplicate(
+            app=request.app,
+            user_id=user_id,
+            exported_project_path=exported_project_path,
+        )
+        return dict(uuid=duplicated_project_uuid)
+    except Exception as e:
+        # make sure all errors are trapped and the directory where the file is sotred is removed
+        await remove_dir(temp_dir)
```
[the remark this answers is not in the document]
Comment (GitHK) [reply services/web/server/src/simcore_service_webserver/exporter/request_handlers.py:118]: you are correct, this was refactored

File: services/web/server/tests/unit/isolated/test_exporter_archiving.py [hunk 6]
```diff
@@ -52,7 +60,18 @@ def returncode(self):
 @pytest.fixture
 def temp_dir() -> Path:
     with tempfile.TemporaryDirectory() as dir_path:
-        yield Path(dir_path)
+        subdir = Path(dir_path) / "subdir_in_temp_dir"
```
[the remark this answers is not in the document]
Comment (GitHK) [reply services/web/server/tests/unit/isolated/test_exporter_archiving.py:63]: But I want to have a subdirectory inside my temporary directory. That's the point of the entire feature.

[the remark this answers is not in the document]
Comment (GitHK) [reply services/web/server/tests/unit/isolated/test_exporter_archiving.py:63]: I see

File: services/web/server/tests/integration/test_exporter.py (hunk 2 above)
Comment (GitHK) [reply services/web/server/tests/integration/test_exporter.py:360]: I see, but will not use in this case., will keep it like this, I find more readable.

File: services/web/server/tests/integration/test_exporter.py (hunk 3 above)
Comment (GitHK) [reply services/web/server/tests/integration/test_exporter.py:301]: You are correct, this should be checked as well. I will compare that all imported files are the same (after being downloaded)

**Review by odeimaiz — APPROVED** [review]

**Review by pcrespov — APPROVED** [review]

**Pull request merged**